nixos/modules/services/hardware/nvidia-container-toolkit.nix

```diff
@@ -0,0 +1,84 @@
+{ config, lib, pkgs, ... }:
+
+{
+
+  options = {
+
+    hardware.nvidia-container-toolkit.enable = lib.mkOption {
+      default = false;
+      type = lib.types.bool;
+      description = lib.mdDoc ''
+        Enable the NVIDIA Container Toolkit (nvidia-ctk) service that
+        generates a Container Device Interface (CDI) configuration file
+        for auto-detected NVIDIA devices, so that CDI-compliant container
+        runtimes can use the forwarded host GPU's.
+      '';
+    };
+
+  };
+
+  config = let
+    generatedCdiConfigPlacement = "/var/run/cdi/nvidia-container-toolkit.json";
+  in {
+    systemd.services.nvidia-container-toolkit = lib.mkIf config.hardware.nvidia-container-toolkit.enable {
+      description = "Container Device Interface (CDI) for Nvidia generator";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "systemd-udev-settle.service" ];
+      serviceConfig = {
+        ExecStart = let
+          script = pkgs.writeScriptBin "nvidia-cdi-generator" ''
+            #! ${pkgs.runtimeShell}
+            mkdir -p /var/run/cdi
+            ${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk cdi generate \
+              --format json \
+              --ldconfig-path ${pkgs.glibc.bin}/bin/ldconfig \
+              --library-search-path ${config.hardware.nvidia.package}/lib \
+              --nvidia-ctk-path ${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-cuda-mps-control", "containerPath": "/usr/bin/nvidia-cuda-mps-control", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
```

Comment:
RE: readability

This makes me think of two things:
1. I'd maybe like to be able to see this in a separate `.sh` file with the proper syntax highlighting, and such that it's easy to pass through `shellcheck`
2. OTOH, the bash script wouldn't be readable anyway, and at the nix level you could get rid of the `\\n` pattern and instead put the rules in a list to `concatStringsSep` later
3. There's a chance a python script with `import json` would still have been more comprehensible

Reply:
I certainly would like to improve the readability of that script. I'll have a look if I can come up with something that is more readable.

Reply:
Updated with a more readable version. WDYT?